src/peer-store/metadata-book.js

```diff
@@ -0,0 +1,165 @@
+'use strict'
+
+const errcode = require('err-code')
+const debug = require('debug')
+const log = debug('libp2p:peer-store:proto-book')
+log.error = debug('libp2p:peer-store:proto-book:error')
+
+const { Buffer } = require('buffer')
+
+const PeerId = require('peer-id')
+
+const Book = require('./book')
+
+const {
+  codes: { ERR_INVALID_PARAMETERS }
+} = require('../errors')
+
+/**
+ * The MetadataBook is responsible for keeping the known supported
+ * protocols of a peer.
+ * @fires MetadataBook#change:metadata
+ */
+class MetadataBook extends Book {
+  /**
+  * @constructor
+  * @param {PeerStore} peerStore
+  */
+  constructor (peerStore) {
+    /**
+     * PeerStore Event emitter, used by the MetadataBook to emit:
+     * "change:metadata" - emitted when the known metadata of a peer change.
+     */
+    super({
+      peerStore,
+      eventName: 'change:metadata',
+      eventProperty: 'metadata'
+    })
+
+    /**
+     * Map known peers to their known protocols.
+     * @type {Map<string, Map<string, Buffer>>}
+     */
+    this.data = new Map()
+  }
+
+  /**
+   * Set metadata key and value of a provided peer.
+   * @override
+   * @param {PeerId} peerId
+   * @param {string} key metadata key
+   * @param {Buffer} value metadata value
+   * @returns {ProtoBook}
+   */
+  set (peerId, key, value) {
+    if (!PeerId.isPeerId(peerId)) {
+      log.error('peerId must be an instance of peer-id to store data')
+      throw errcode(new Error('peerId must be an instance of peer-id'), ERR_INVALID_PARAMETERS)
+    }
+
+    if (!key || !value || !(typeof key === 'string') || !Buffer.isBuffer(value)) {
```

Comment:
This can be simplified. We could potentially just do a `Buffer.from` try/catch to support strings/arraybuffer/etc out of the box as well.
```suggestion
    if (typeof key !== 'string') || !Buffer.isBuffer(value)) {
```

Reply:
I wanted to go with the Buffer only support because if a user sets metadata with a string, it will work, and then on get a. buffer will be returned... The code suggestion seems good to me